File: backend/service.py
"""
业务服务
合并登录服务和阳光跑服务
"""
import re
from typing import Optional, Tuple

import httpx

from core.config import BASE_URL, WECHAT_APPID, WECHAT_BUNDLE_ID, COMMON_HEADERS, WECHAT_HEADERS
from core.crypto import rsa_encrypt
from models import LoginInfo, QRCodeInfo, RunTask
from utils import generate_run_data


# ============================================================
# 登录服务
# ============================================================

class LoginService:
    """微信扫码登录服务"""

    def __init__(self):
        self.client = httpx.AsyncClient(timeout=30.0)

    async def close(self):
        await self.client.aclose()

    async def __aenter__(self):
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        await self.close()

    async def get_qrcode(self) -> QRCodeInfo:
        """获取微信登录二维码"""
        url = (
            f"https://open.weixin.qq.com/connect/app/qrconnect"
            f"?appid={WECHAT_APPID}"
            f"&bundleid={WECHAT_BUNDLE_ID}"
            f"&scope=snsapi_userinfo"
            f"&state="
        )

        response = await self.client.get(url, headers=WECHAT_HEADERS)
        response.raise_for_status()

        html = response.text

        uuid_match = re.search(r'uuid\s*=\s*["\']([^"\']+)["\']', html)
        if not uuid_match:
            uuid_match = re.search(r'/qrcode/([a-zA-Z0-9_-]+)', html)

        if not uuid_match:
            raise ValueError("无法从页面中提取 UUID")

        uuid = uuid_match.group(1)
        qrcode_url = f"https://open.weixin.qq.com/connect/qrcode/{uuid}"

        return QRCodeInfo(qrcode_url=qrcode_url, uuid=uuid)

    async def poll_scan_status(self, uuid: str) -> Tuple[int, str, Optional[str]]:
        """
        轮询一次扫码状态

        Returns:
            (状态码, 状态消息, 微信授权码或None)
            状态码: 408=等待扫码, 404=已扫码, 405=已授权, 403=取消, 402=过期
        """
        poll_url = f"https://long.open.weixin.qq.com/connect/l/qrconnect?uuid={uuid}&f=json"

        try:
            response = await self.client.get(poll_url, headers=WECHAT_HEADERS, timeout=30.0)
            response.raise_for_status()

            data = response.json()
            wx_errcode = data.get("wx_errcode", 408)
            wx_code = data.get("wx_code", "")

            status_messages = {
                408: "等待扫码",
                404: "已扫码，请在手机上确认",
                405: "授权成功",
                403: "用户取消授权",
                402: "二维码已过期"
            }

            message = status_messages.get(wx_errcode, f"未知状态: {wx_errcode}")
            return wx_errcode, message, wx_code if wx_code else None

        except httpx.RequestError as e:
            return 408, f"请求错误: {str(e)}", None

    async def get_server_token(self, wechat_code: str) -> str:
        """使用微信授权码获取服务器 token"""
        url = f"{BASE_URL}/app/platform/serverlist/getLesseeServer"

        data = {"code": wechat_code}
        encrypted_body = rsa_encrypt(data)

        response = await self.client.post(
            url, content=encrypted_body, headers=COMMON_HEADERS
        )
        response.raise_for_status()

        result = response.json()
        if result.get("code") != "0":
            raise ValueError(f"获取服务器token失败: {result.get('message', '未知错误')}")

        return result.get("token", "")

    async def login(
        self,
        wechat_code: str,
        server_token: str,
        longitude: str = "116.397428",
        latitude: str = "39.908823"
    ) -> LoginInfo:
        """执行登录"""
        url = f"{BASE_URL}/app/platform/login/login"

        data = {
            "loginWay": "1",
            "phoneNumber": "",
            "password": "",
            "code": wechat_code,
            "longitude": longitude,
            "latitude": latitude,
            "token": server_token
        }

        encrypted_body = rsa_encrypt(data)
        response = await self.client.post(
            url, content=encrypted_body, headers=COMMON_HEADERS
        )
        response.raise_for_status()

        result = response.json()
        return LoginInfo.from_dict(result, server_token=server_token)

    async def complete_login(
        self,
        wx_code: str,
        longitude: str = "116.397428",
        latitude: str = "39.908823"
    ) -> LoginInfo:
        """完成登录（获取token + 登录）"""
        server_token = await self.get_server_token(wx_code)
        login_info = await self.login(wx_code, server_token, longitude, latitude)
        return login_info


# 单例
_login_service: Optional[LoginService] = None


async def get_login_service() -> LoginService:
    """获取登录服务单例"""
    global _login_service
    if _login_service is None:
        _login_service = LoginService()
    return _login_service


# ============================================================
# 阳光跑服务
# ============================================================

class SunRunService:
    """阳光跑服务"""

    def __init__(self, token: str, stu_number: str, school_id: str, campus_id: str):
        self.token = token
        self.stu_number = stu_number
        self.school_id = school_id
        self.campus_id = campus_id
        self.client = httpx.AsyncClient(timeout=30.0)

    async def close(self):
        await self.client.aclose()

    async def __aenter__(self):
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        await self.close()

    async def _post(self, endpoint: str, data: dict, encrypt: bool = True) -> dict:
        """发送 POST 请求"""
        url = f"{BASE_URL}{endpoint}"
        data["token"] = self.token

        if encrypt:
            encrypted_body = rsa_encrypt(data)
            response = await self.client.post(url, content=encrypted_body, headers=COMMON_HEADERS)
        else:
            response = await self.client.post(url, json=data, headers=COMMON_HEADERS)

        response.raise_for_status()
        return response.json()

    async def get_sunrun_task(self) -> RunTask:
        """获取阳光跑任务"""
        data = {
            "stuNumber": self.stu_number,
            "schoolId": self.school_id,
            "campusId": self.campus_id,
        }
        result = await self._post("/app/sunrun/getSunrunPaper", data)
        return RunTask.from_dict(result)

    async def start_run(self) -> dict:
        """开始跑步（必须在提交记录前调用）"""
        data = {
            "stuNumber": self.stu_number,
            "schoolId": self.school_id,
            "campusId": self.campus_id,
        }
        return await self._post("/app/sunrun/getRunBegin", data)

    async def get_sunrun_sport(
        self,
        run_type: str = "0",
        month_id: str = "",
        page_number: str = "1",
        row_number: str = "100"
    ) -> dict:
        """获取跑步记录列表"""
        data = {
            "stuNumber": self.stu_number,
            "runType": run_type,
            "monthId": month_id,
            "pageNumber": page_number,
            "rowNumber": row_number,
        }
        return await self._post("/app/sunrun/getSunrunSport", data)

    async def submit_run_record(self, run_data: dict) -> dict:
        """提交跑步基础记录"""
        return await self._post("/app/platform/recrecord/sunRunExercises", run_data)

    async def submit_run_detail(self, scantron_id: str, point_list: list) -> dict:
        """提交跑步详情（轨迹点）— 不加密"""
        data = {
            "scantronId": scantron_id,
            "stuNumber": self.stu_number,
            "faceData": "",
            "pointList": point_list,
            "token": self.token
        }

        url = f"{BASE_URL}/app/platform/recrecord/sunRunExercisesDetail"
        response = await self.client.post(url, json=data, headers=COMMON_HEADERS)
        response.raise_for_status()
        return response.json()

    async def submit_complete_run(
        self,
        task: RunTask,
        campus_name: str,
        km: float = None,
        used_time_minutes: int = None,
        point_index: int = None,
        run_date: str = None,
        run_time: str = None
    ) -> tuple:
        """完整提交跑步记录（基础记录 + 轨迹详情）"""
        # 1. 生成跑步数据
        run_data = generate_run_data(
            task=task,
            stu_number=self.stu_number,
            token=self.token,
            campus_name=campus_name,
            km=km,
            used_time_minutes=used_time_minutes,
            point_index=point_index,
            run_date=run_date,
            run_time=run_time
        )

        point_list = run_data.pop("_pointList", [])
        selected_point = run_data.pop("_selectedPoint", "默认")

        print(f"[service] 生成的跑步数据:")
        print(f"  submitDate={run_data.get('submitDate')}")
        print(f"  evaluateDate={run_data.get('evaluateDate')}")
        print(f"  startTime={run_data.get('startTime')}")
        print(f"  endTime={run_data.get('endTime')}")
        print(f"  km={run_data.get('km')}, usedTime={run_data.get('usedTime')}")
        print(f"  ifLocalSubmit={run_data.get('ifLocalSubmit')}")
        print(f"  selectedPoint={selected_point}")

        # 2. 提交基础记录
        result1 = await self.submit_run_record(run_data)

        if result1.get("code") != "0":
            return result1, None, {
                "km": run_data["km"],
                "usedTime": run_data["usedTime"],
                "selectedPoint": selected_point,
                "trackPoints": len(point_list)
            }

        scantron_id = result1.get("scantronId", "")

        # 3. 提交轨迹详情
        result2 = await self.submit_run_detail(scantron_id, point_list)

        return result1, result2, {
            "km": run_data["km"],
            "usedTime": run_data["usedTime"],
            "avgSpeed": run_data["avgSpeed"],
            "steps": run_data["steps"],
            "selectedPoint": selected_point,
            "trackPoints": len(point_list),
            "scantronId": scantron_id
        }

File: lib/server/service.js
import {
  BASE_URL,
  COMMON_HEADERS,
  WECHAT_APPID,
  WECHAT_BUNDLE_ID,
  WECHAT_HEADERS,
} from './config';
import { rsaEncrypt } from './crypto';
import { makeLoginInfo, makeRunTask } from './models';
import { generateRunData } from './run-data';

async function fetchJson(url, options = {}, timeoutMs = 30000) {
  const response = await fetch(url, {
    ...options,
    cache: 'no-store',
    signal: AbortSignal.timeout(timeoutMs),
  });

  if (!response.ok) {
    const text = await response.text().catch(() => '');
    console.error('[upstream] non-ok response', {
      url,
      status: response.status,
      body: text.slice(0, 500),
    });
    throw new Error(`上游请求失败: ${response.status}${text ? ` ${text.slice(0, 120)}` : ''}`);
  }

  try {
    return await response.json();
  } catch (error) {
    const text = await response.text().catch(() => '');
    console.error('[upstream] invalid json response', {
      url,
      message: error.message,
      body: text.slice(0, 500),
    });
    throw new Error(`上游响应不是JSON: ${error.message}`);
  }
}

export class LoginService {
  async getQRCode() {
    const url =
      `https://open.weixin.qq.com/connect/app/qrconnect?appid=${WECHAT_APPID}` +
      `&bundleid=${WECHAT_BUNDLE_ID}&scope=snsapi_userinfo&state=`;
    const response = await fetch(url, { headers: WECHAT_HEADERS, signal: AbortSignal.timeout(10000) });

    if (!response.ok) {
      throw new Error(`获取二维码失败: ${response.status}`);
    }

    const html = await response.text();
    const uuidMatch = html.match(/uuid\s*=\s*["']([^"']+)["']/) || html.match(/\/qrcode\/([a-zA-Z0-9_-]+)/);

    if (!uuidMatch) {
      throw new Error('无法从页面中提取 UUID');
    }

    const uuid = uuidMatch[1];
    return {
      qrcode_url: `https://open.weixin.qq.com/connect/qrcode/${uuid}`,
      uuid,
    };
  }

  async pollScanStatus(uuid) {
    const url = `https://long.open.weixin.qq.com/connect/l/qrconnect?uuid=${uuid}&f=json`;

    try {
      const data = await fetchJson(url, { headers: WECHAT_HEADERS }, 8000);
      const wxErrcode = data.wx_errcode ?? 408;
      const wxCode = data.wx_code || '';
      const messages = {
        408: '等待扫码',
        404: '已扫码，请在手机上确认',
        405: '授权成功',
        403: '用户取消授权',
        402: '二维码已过期',
      };

      return {
        status: wxErrcode,
        message: messages[wxErrcode] || `未知状态: ${wxErrcode}`,
        wx_code: wxCode || null,
      };
    } catch {
      return {
        status: 408,
        message: '等待扫码',
        wx_code: null,
      };
    }
  }

  async getServerToken(wechatCode) {
    const body = rsaEncrypt({ code: wechatCode });
    const result = await fetchJson(`${BASE_URL}/app/platform/serverlist/getLesseeServer`, {
      method: 'POST',
      body,
      headers: COMMON_HEADERS,
    });

    if (result.code !== '0') {
      console.error('[login] getServerToken failed', {
        code: result.code,
        message: result.message,
      });
      throw new Error(`获取服务器token失败: ${result.message || '未知错误'}`);
    }

    return result.token || '';
  }

  async login(wechatCode, serverToken, longitude = '116.397428', latitude = '39.908823') {
    const body = rsaEncrypt({
      loginWay: '1',
      phoneNumber: '',
      password: '',
      code: wechatCode,
      longitude,
      latitude,
      token: serverToken,
    });
    const result = await fetchJson(`${BASE_URL}/app/platform/login/login`, {
      method: 'POST',
      body,
      headers: COMMON_HEADERS,
    });

    return makeLoginInfo(result, serverToken);
  }

  async completeLogin(wxCode, longitude, latitude) {
    const serverToken = await this.getServerToken(wxCode);
    return this.login(wxCode, serverToken, longitude, latitude);
  }
}

export class SunRunService {
  constructor({ token, stuNumber, schoolId, campusId }) {
    this.token = token;
    this.stuNumber = stuNumber;
    this.schoolId = schoolId;
    this.campusId = campusId;
  }

  async post(endpoint, data, encrypt = true) {
    const payload = { ...data, token: this.token };
    const options = encrypt
      ? { body: rsaEncrypt(payload), headers: COMMON_HEADERS }
      : { body: JSON.stringify(payload), headers: COMMON_HEADERS };

    return fetchJson(`${BASE_URL}${endpoint}`, {
      method: 'POST',
      ...options,
    });
  }

  async getSunrunTask() {
    const result = await this.post('/app/sunrun/getSunrunPaper', {
      stuNumber: this.stuNumber,
      schoolId: this.schoolId,
      campusId: this.campusId,
    });
    return makeRunTask(result);
  }

  async startRun() {
    return this.post('/app/sunrun/getRunBegin', {
      stuNumber: this.stuNumber,
      schoolId: this.schoolId,
      campusId: this.campusId,
    });
  }

  async getSunrunSport({ runType = '0', monthId = '', pageNumber = '1', rowNumber = '100' } = {}) {
    return this.post('/app/sunrun/getSunrunSport', {
      stuNumber: this.stuNumber,
      runType,
      monthId,
      pageNumber,
      rowNumber,
    });
  }

  async submitRunRecord(runData) {
    return this.post('/app/platform/recrecord/sunRunExercises', runData);
  }

  async submitRunDetail(scantronId, pointList) {
    return this.post(
      '/app/platform/recrecord/sunRunExercisesDetail',
      {
        scantronId,
        stuNumber: this.stuNumber,
        faceData: '',
        pointList,
      },
      false
    );
  }

  async submitCompleteRun({ task, campusName, km, usedTimeMinutes, pointIndex, runDate, runTime }) {
    const runData = generateRunData({
      task,
      stuNumber: this.stuNumber,
      token: this.token,
      campusName,
      km,
      usedTimeMinutes,
      pointIndex,
      runDate,
      runTime,
    });
    const pointList = runData._pointList || [];
    const selectedPoint = runData._selectedPoint || '默认';
    delete runData._pointList;
    delete runData._selectedPoint;

    const result1 = await this.submitRunRecord(runData);

    if (result1.code !== '0') {
      return {
        result1,
        result2: null,
        runInfo: {
          km: runData.km,
          usedTime: runData.usedTime,
          selectedPoint,
          trackPoints: pointList.length,
        },
      };
    }

    const scantronId = result1.scantronId || '';
    const result2 = await this.submitRunDetail(scantronId, pointList);

    return {
      result1,
      result2,
      runInfo: {
        km: runData.km,
        usedTime: runData.usedTime,
        avgSpeed: runData.avgSpeed,
        steps: runData.steps,
        selectedPoint,
        trackPoints: pointList.length,
        scantronId,
      },
    };
  }
}
